src/libmongoc/src/mongoc/mongoc-client-pool.c

```diff
@@ -357,6 +363,76 @@ mongoc_client_pool_try_pop (mongoc_client_pool_t *pool)
    RETURN (client);
 }
 
+typedef struct {
+   mongoc_array_t *server_ids;
+   mongoc_cluster_t *cluster;
+} prune_ctx;
+
+static int
+server_id_cmp (const void *a_, const void *b_)
+{
+   const uint32_t *const a = (const uint32_t *) a_;
+   const uint32_t *const b = (const uint32_t *) b_;
+
+   if (*a == *b) {
+      return 0;
+   }
+
+   return *a < *b ? -1 : 1;
+}
+
+static bool
+maybe_prune (void *item, void *ctx_)
+{
+   mongoc_cluster_node_t *cn = (mongoc_cluster_node_t *) item;
+   prune_ctx *ctx = (prune_ctx *) ctx_;
+   uint32_t server_id = cn->handshake_sd->id;
+
+   if (!bsearch (&server_id, ctx->server_ids->data, ctx->server_ids->len, sizeof (uint32_t), server_id_cmp)) {
+      mongoc_cluster_disconnect_node (ctx->cluster, server_id);
+   }
+   return true;
+}
+
+// `prune_client` closes connections from `client` to servers not contained in `known_server_ids`.
+static void
+prune_client (mongoc_client_t *client, mongoc_array_t *known_server_ids)
+{
+   BSON_ASSERT_PARAM (client);
+   BSON_ASSERT_PARAM (known_server_ids);
+
+   mongoc_cluster_t *cluster = &client->cluster;
+
+   bool needs_prune = false;
+   // Do a fast initial check to see if a prune is needed.
+   size_t idx1 = 0;
+   size_t idx2 = 0;
+   for (; idx1 < cluster->nodes->items_len; idx1++) {
+      // Compare both sorted lists in order. `cluster->nodes` may be a smaller list if not all servers were used.
+      mongoc_set_item_t *cn = &cluster->nodes->items[idx1];
+      bool found = false;
+      for (; idx2 < known_server_ids->len; idx2++) {
+         uint32_t last_known = _mongoc_array_index (known_server_ids, uint32_t, idx2);
+         if (cn->id == last_known) {
+            found = true;
+            break;
+         }
+      }
+      if (!found) {
+         // A server in the cluster is not in the last known server ids. Prune it.
+         needs_prune = true;
+         break;
+      }
+   }
+
+   if (!needs_prune) {
+      return;
+   }
```

Comment:
Is it likely that this early-return saves a significant amount of time for the additional complexity that it adds?

Reply:
Good question. Another benchmark does not show significant change with [this early-return removed](https://spruce.mongodb.com/task/mongo_c_driver_perf_c_driver_benchmark_mongo60_perf_BenchMarkTests_patch_9fe27e912525ca127ef889660cb9ebfa802bbe5c_665871d1450b780007ad9fde_24_05_30_12_32_18/logs?execution=0):


| Test                                   |   Baseline (all conditional pruning) | Without conditional client pruning   | Without conditional pruning (simple prune)   |
|----------------------------------------|--------------------------------------|--------------------------------------|----------------------------------------------|
| Parallel/Pool/ShareClients/Threads:1   |                               249159 | 246554 (-1.05%)                      | 235143 (-5.63%)                              |
| Parallel/Pool/ShareClients/Threads:10  |                              1312571 | 1313206 (0.05%)                      | 866941 (-33.95%)                             |
| Parallel/Pool/ShareClients/Threads:100 |                              1932369 | 1917569 (-0.77%)                     | 1194404 (-38.19%)                            |
| Parallel/Pool/ShareClients/Threads:150 |                              1555300 | 1564356 (0.58%)                      | 1564161 (0.57%)                              |

The "Baseline" is the current changeset. The "Without conditional client pruning" removes this early-return.

Updated to remove the early-return.